posts/2023-07-18-lazy-layout.lhs

```diff
@@ -0,0 +1,589 @@
+the photographer envisioned it: concretely, we can scale it up or down
+(preserving the aspect ratio), but we can't crop out parts.
+
+Order is also important in photo essays, so we want the author to specify the
+photo collage in a declarative way by indicating if horizontal (H) or vertical
+(V) subdivision should be used, creating a tree.  For example:
+
+    H img1.jpg
+      (V img2.jpg
+         (H img3.jpg
+            img4.jpg))
+
+The program should then determine the exact size and position of each image,
+so that we get a fully filled rectangle without any borders or filler:
+
+![](/images/2023-07-18-lazy-layout-example-1.jpg)
+
+We will use a technique called _circular programming_ that builds on Haskell's
+laziness to achieve this in an elegant way.
+These days, it is maybe more commonly referred to as the `repmin` problem.
+This was first described by Richard S. Bird in _"Using circular programs to
+eliminate multiple traversals of data"_ in 1984, which **predates Haskell!**
+
+<details><summary>Give me a refresher on <code>repmin</code> please...</summary>
+
+Interlude: repmin
+-----------------
+
+Given a simple tree type:
+
+> data Tree a
+>   = Leaf a
+>   | Branch (Tree a) (Tree a)
+
+We would like to write a function `repmin` which replaces each value in each
+`Leaf` with the global minimum in the tree.  This is easily done by first
+finding the global minimum, and then replacing it everywhere:
+
+> repmin_2pass :: Ord a => Tree a -> Tree a
+> repmin_2pass t =
+>   let globalmin = findmin t in rep globalmin t
+>  where
+>   findmin (Leaf x)     = x
+>   findmin (Branch l r) = min (findmin l) (findmin r)
+>
+>   rep x (Leaf _)     = Leaf x
+>   rep x (Branch l r) = Branch (rep x l) (rep x r)
+
+However, this requires two passes over the tree.  We can do better by using
+Haskell's laziness:
+
+> repmin_1pass :: Ord a => Tree a -> Tree a
+> repmin_1pass t = t'
+>  where
+>   (t', globalmin) = repmin t
+>
+>   repmin (Leaf   x)   = (Leaf globalmin, x)
+>   repmin (Branch l r) =
+>     (Branch l' r', min lmin rmin)
+>    where
+>     (l', lmin) = repmin l
+>     (r', rmin) = repmin r
+
+There is an apparent circular dependency here, where `repmin` uses `globalmin`,
+but also computes it.  This is possible because we never need to evaluate
+`globalmin` -- it can be stored as a [thunk].
+For more details, please see the very accessible original paper
+(`https://doi.org/10.1007/BF00264249`).
+
+[thunk]: https://www.youtube.com/watch?v=I4lnCG18TaY
+
+</details>
+
+Starting out with some types
+============================
+
+We start out simple by giving an elegant algebraic definition for a collage:
+
+> data Collage a
+>   = Singleton  a
+>   | Horizontal (Collage a) (Collage a)
+>   | Vertical   (Collage a) (Collage a)
+>   deriving (Foldable, Functor, Show, Traversable)
+
+We will use the [JuicyPixels] library to load and write images.
+The image type in this library can be a bit verbose since it is parameterized
```

Comment:
Seeing how you write `colour` I assume you're using British English. In that case:

```suggestion
The image type in this library can be a bit verbose since it is parameterised
```